File: skills/twg-jira/scripts/jira-workitem-implementation.mjs
#!/usr/bin/env node

import { spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { homedir } from "node:os";
import path from "node:path";
import { pathToFileURL } from "node:url";

export const JIRA_FIELDS = "summary,description,status,parent,issuetype";

function contentOf(node) {
  return Array.isArray(node?.content) ? node.content : [];
}

export function renderInlines(nodes = []) {
  return nodes.map((node) => renderInline(node)).join("");
}

export function renderInline(node) {
  if (node?.type === "text") {
    let text = typeof node.text === "string" ? node.text : "";
    for (const mark of Array.isArray(node.marks) ? node.marks : []) {
      if (mark.type === "code") text = `\`${text}\``;
      else if (mark.type === "strong") text = `**${text}**`;
      else if (mark.type === "em") text = `*${text}*`;
      else if (mark.type === "strike") text = `~~${text}~~`;
      else if (mark.type === "link" && mark.attrs?.href)
        text = `[${text}](${String(mark.attrs.href)})`;
    }
    return text;
  }
  if (node?.type === "hardBreak") return "\n";
  if (["mention", "emoji", "status"].includes(node?.type)) {
    return String(
      node.attrs?.text ??
        node.attrs?.displayName ??
        node.attrs?.shortName ??
        "",
    );
  }
  if (["inlineCard", "blockCard", "embedCard"].includes(node?.type))
    return String(node.attrs?.url ?? "");
  return renderInlines(contentOf(node));
}

export function renderBlock(node) {
  const content = contentOf(node);
  if (node?.type === "doc")
    return content
      .map((child) => renderBlock(child).trim())
      .filter(Boolean)
      .join("\n\n");
  if (node?.type === "heading") {
    const level = Math.min(Math.max(Number(node.attrs?.level) || 1, 1), 6);
    return `${"#".repeat(level)} ${renderInlines(content)}`;
  }
  if (node?.type === "paragraph") return renderInlines(content);
  if (node?.type === "codeBlock") {
    const language = String(node.attrs?.language ?? "");
    return `\`\`\`${language}\n${renderInlines(content)}\n\`\`\``;
  }
  if (node?.type === "blockquote") {
    const rendered = content.map((child) => renderBlock(child)).join("\n\n");
    return rendered
      .split(/\r\n|\r|\n/u)
      .map((line) => (line ? `> ${line}` : ">"))
      .join("\n");
  }
  if (node?.type === "rule") return "---";
  if (["bulletList", "orderedList"].includes(node?.type)) {
    const start = Number(node.attrs?.order) || 1;
    return content
      .map((item, index) => {
        const marker = node.type === "bulletList" ? "-" : `${start + index}.`;
        const lines = renderBlock(item).trim().split(/\r\n|\r|\n/u);
        return [
          `${marker} ${lines[0] ?? ""}`,
          ...lines.slice(1).map((line) => `  ${line}`),
        ].join("\n");
      })
      .join("\n");
  }
  if (node?.type === "listItem")
    return content
      .map((child) => renderBlock(child).trim())
      .filter(Boolean)
      .join("\n");
  if (node?.type === "table") {
    const rows = content.map((row) => renderTableRow(row));
    if (rows.length === 0) return "";
    const separator = rows[0].map(() => "---");
    return [rows[0], separator, ...rows.slice(1)]
      .map((row) => renderTableCells(row))
      .join("\n");
  }
  if (["tableCell", "tableHeader"].includes(node?.type))
    return content
      .map((child) => renderBlock(child).trim())
      .filter(Boolean)
      .join(" ");
  if (["panel", "expand", "nestedExpand"].includes(node?.type)) {
    const title = typeof node.attrs?.title === "string" ? node.attrs.title : "";
    const body = content
      .map((child) => renderBlock(child).trim())
      .filter(Boolean)
      .join("\n\n");
    return [title, body].filter(Boolean).join("\n\n");
  }
  if (content.length > 0)
    return content
      .map((child) => renderBlock(child).trim())
      .filter(Boolean)
      .join("\n\n");
  return renderInline(node);
}

function renderTableRow(row) {
  return contentOf(row).map((cell) =>
    renderBlock(cell).replaceAll("|", "\\|").replaceAll(/\r\n|\r|\n/gu, "<br>"),
  );
}

function renderTableCells(cells) {
  return `| ${cells.join(" | ")} |`;
}

function compactText(value) {
  if (value?.type === "doc") return renderBlock(value).trim();
  return typeof value === "string" ? value.trim() : undefined;
}

function namedValue(value) {
  if (value && typeof value === "object") return value.name;
  return typeof value === "string" ? value : undefined;
}

function compactObject(value) {
  return Object.fromEntries(
    Object.entries(value).filter(
      ([, item]) =>
        item !== undefined &&
        item !== null &&
        item !== "" &&
        !(Array.isArray(item) && item.length === 0) &&
        !(
          typeof item === "object" &&
          !Array.isArray(item) &&
          Object.keys(item).length === 0
        ),
    ),
  );
}

export function projectWorkitem(workitem) {
  const parent = workitem?.parent;
  const parentFields =
    parent?.fields && typeof parent.fields === "object" ? parent.fields : {};
  const projectedParent =
    parent && typeof parent === "object"
      ? compactObject({
          key: parent.key,
          summary: parent.summary ?? parentFields.summary,
        })
      : undefined;
  return compactObject({
    key: workitem?.key,
    summary: workitem?.summary,
    status: namedValue(workitem?.status),
    issueType: namedValue(workitem?.issuetype),
    parent: projectedParent,
    description: compactText(workitem?.description),
    url: workitem?.url,
  });
}

export function normalizePayload(payload) {
  if (Array.isArray(payload)) return payload;
  if (!payload || typeof payload !== "object")
    throw new TypeError("Unexpected TWG Jira payload shape");
  const data = payload.data ?? payload;
  if (Array.isArray(data)) return data;
  if (Array.isArray(data?.items))
    return data.items.map((item) => item?.data ?? item);
  if (data && typeof data === "object") return [data];
  throw new TypeError("Unexpected TWG Jira payload shape");
}

export function buildArguments(keys) {
  return [
    "jira",
    "workitem",
    "get",
    ...keys,
    "--fields",
    JIRA_FIELDS,
    "-o",
    "json",
    "--output-summary",
    "none",
  ];
}

function twgCandidates() {
  const fallback =
    process.platform === "win32" && process.env.LOCALAPPDATA
      ? path.join(
          process.env.LOCALAPPDATA,
          "Programs",
          "twg",
          "bin",
          "twg.exe",
        )
      : path.join(homedir(), ".local", "bin", "twg");
  return ["twg", fallback];
}

function fetchWorkitems(keys) {
  for (const binary of twgCandidates()) {
    if (binary !== "twg" && !existsSync(binary)) continue;
    const result = spawnSync(binary, buildArguments(keys), {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"],
    });
    if (result.error?.code === "ENOENT") continue;
    if (result.error) throw result.error;
    if (result.status !== 0)
      throw new Error(
        result.stderr.trim() || "TWG command failed without diagnostic output",
      );
    try {
      return JSON.parse(result.stdout);
    } catch (error) {
      throw new Error("TWG did not return JSON", { cause: error });
    }
  }
  throw new Error("Required command not found: twg");
}

function parseArguments(arguments_) {
  const keys = [];
  let inputPath;
  for (let index = 0; index < arguments_.length; index += 1) {
    const argument = arguments_[index];
    if (argument === "--input") {
      inputPath = arguments_[index + 1];
      if (!inputPath || inputPath.startsWith("-"))
        throw new Error("--input requires a path");
      index += 1;
    } else if (argument === "-h" || argument === "--help") {
      return { help: true, keys: [] };
    } else if (argument.startsWith("-")) {
      throw new Error(`Unknown option: ${argument}`);
    } else {
      keys.push(argument);
    }
  }
  if (inputPath && keys.length > 0)
    throw new Error("Jira keys cannot be combined with --input");
  if (!inputPath && keys.length === 0)
    throw new Error("Provide at least one Jira key or --input");
  return { help: false, inputPath, keys };
}

export async function run(arguments_ = process.argv.slice(2)) {
  try {
    const options = parseArguments(arguments_);
    if (options.help) {
      process.stdout.write(
        "Usage: jira-workitem-implementation.mjs <KEY...> | --input <payload.json>\n",
      );
      return 0;
    }
    const payload = options.inputPath
      ? JSON.parse(await readFile(options.inputPath, "utf8"))
      : fetchWorkitems(options.keys);
    const items = normalizePayload(payload).map((item) =>
      projectWorkitem(item),
    );
    process.stdout.write(`${JSON.stringify({ items }, null, 2)}\n`);
    return 0;
  } catch (error) {
    process.stderr.write(
      `jira-workitem-implementation: ${
        error instanceof Error ? error.message : String(error)
      }\n`,
    );
    return 1;
  }
}

const invokedPath = process.argv[1];
if (
  invokedPath &&
  import.meta.url === pathToFileURL(path.resolve(invokedPath)).href
) {
  process.exitCode = await run();
}

File: skills/twg-jira/scripts/jira_workitem_implementation.py
#!/usr/bin/env python3

import argparse
import json
import shutil
import subprocess
import sys
from pathlib import Path


JIRA_FIELDS = "summary,description,status,parent,issuetype"


def render_inlines(nodes):
    return "".join(render_inline(node) for node in nodes or [])


def render_inline(node):
    node_type = node.get("type")
    if node_type == "text":
        text = node.get("text", "")
        for mark in node.get("marks", []):
            mark_type = mark.get("type")
            if mark_type == "code":
                text = f"`{text}`"
            elif mark_type == "strong":
                text = f"**{text}**"
            elif mark_type == "em":
                text = f"*{text}*"
            elif mark_type == "strike":
                text = f"~~{text}~~"
            elif mark_type == "link":
                href = mark.get("attrs", {}).get("href")
                if href:
                    text = f"[{text}]({href})"
        return text
    if node_type == "hardBreak":
        return "\n"
    if node_type in {"mention", "emoji", "status"}:
        attributes = node.get("attrs", {})
        return (
            attributes.get("text")
            or attributes.get("displayName")
            or attributes.get("shortName")
            or ""
        )
    if node_type in {"inlineCard", "blockCard", "embedCard"}:
        return node.get("attrs", {}).get("url", "")
    return render_inlines(node.get("content"))


def render_block(node):
    node_type = node.get("type")
    content = node.get("content", [])
    if node_type == "doc":
        return "\n\n".join(filter(None, (render_block(child).strip() for child in content)))
    if node_type == "heading":
        level = min(max(int(node.get("attrs", {}).get("level", 1)), 1), 6)
        return f"{'#' * level} {render_inlines(content)}"
    if node_type == "paragraph":
        return render_inlines(content)
    if node_type == "codeBlock":
        language = node.get("attrs", {}).get("language", "")
        return f"```{language}\n{render_inlines(content)}\n```"
    if node_type == "blockquote":
        rendered = "\n\n".join(render_block(child) for child in content)
        return "\n".join(f"> {line}" if line else ">" for line in rendered.splitlines())
    if node_type == "rule":
        return "---"
    if node_type in {"bulletList", "orderedList"}:
        start = int(node.get("attrs", {}).get("order", 1))
        rendered_items = []
        for index, item in enumerate(content):
            marker = "-" if node_type == "bulletList" else f"{start + index}."
            body = render_block(item).strip()
            lines = body.splitlines() or [""]
            rendered_items.append(
                "\n".join([f"{marker} {lines[0]}", *(f"  {line}" for line in lines[1:])])
            )
        return "\n".join(rendered_items)
    if node_type == "listItem":
        return "\n".join(filter(None, (render_block(child).strip() for child in content)))
    if node_type == "table":
        rows = [render_table_row(row) for row in content]
        if not rows:
            return ""
        separator = "| " + " | ".join("---" for _ in rows[0]) + " |"
        rendered = [render_table_cells(rows[0]), separator]
        rendered.extend(render_table_cells(row) for row in rows[1:])
        return "\n".join(rendered)
    if node_type in {"tableCell", "tableHeader"}:
        return " ".join(filter(None, (render_block(child).strip() for child in content)))
    if node_type in {"panel", "expand", "nestedExpand"}:
        title = node.get("attrs", {}).get("title")
        body = "\n\n".join(filter(None, (render_block(child).strip() for child in content)))
        return "\n\n".join(filter(None, (title, body)))
    if content:
        return "\n\n".join(filter(None, (render_block(child).strip() for child in content)))
    return render_inline(node)


def render_table_row(row):
    return [render_block(cell).replace("|", "\\|").replace("\n", "<br>") for cell in row.get("content", [])]


def render_table_cells(cells):
    return "| " + " | ".join(cells) + " |"


def compact_text(value):
    if isinstance(value, dict) and value.get("type") == "doc":
        return render_block(value).strip()
    if isinstance(value, str):
        return value.strip()
    return None


def named_value(value):
    if isinstance(value, dict):
        return value.get("name")
    return value if isinstance(value, str) else None


def project_workitem(workitem):
    parent = workitem.get("parent")
    projected_parent = None
    if isinstance(parent, dict):
        parent_fields = parent.get("fields") if isinstance(parent.get("fields"), dict) else {}
        projected_parent = compact_object(
            {
                "key": parent.get("key"),
                "summary": parent.get("summary") or parent_fields.get("summary"),
            }
        )
    return compact_object(
        {
            "key": workitem.get("key"),
            "summary": workitem.get("summary"),
            "status": named_value(workitem.get("status")),
            "issueType": named_value(workitem.get("issuetype")),
            "parent": projected_parent,
            "description": compact_text(workitem.get("description")),
            "url": workitem.get("url"),
        }
    )


def compact_object(value):
    return {key: item for key, item in value.items() if item not in (None, "", {}, [])}


def normalize_payload(payload):
    if isinstance(payload, list):
        return payload
    if isinstance(payload, dict):
        data = payload.get("data", payload)
        if isinstance(data, list):
            return data
        if isinstance(data, dict) and isinstance(data.get("items"), list):
            return [
                item.get("data", item) if isinstance(item, dict) else item
                for item in data["items"]
            ]
        if isinstance(data, dict):
            return [data]
    raise ValueError("Unexpected TWG Jira payload shape")


def build_command(binary, keys):
    return [
        binary,
        "jira",
        "workitem",
        "get",
        *keys,
        "--fields",
        JIRA_FIELDS,
        "-o",
        "json",
        "--output-summary",
        "none",
    ]


def fetch_workitems(keys):
    binary = shutil.which("twg")
    if binary is None:
        fallback = Path.home() / ".local/bin/twg"
        if not fallback.is_file():
            raise RuntimeError("Required command not found: twg")
        binary = str(fallback)
    result = subprocess.run(
        build_command(binary, keys),
        capture_output=True,
        check=False,
        text=True,
    )
    if result.returncode != 0:
        detail = result.stderr.strip() or "TWG command failed without diagnostic output"
        raise RuntimeError(detail)
    try:
        return json.loads(result.stdout)
    except json.JSONDecodeError as error:
        raise RuntimeError("TWG did not return JSON") from error


def parse_arguments():
    parser = argparse.ArgumentParser(
        description="Fetch Jira workitems and emit implementation-bearing fields with ADF as Markdown."
    )
    parser.add_argument("keys", nargs="*", help="Jira workitem keys")
    parser.add_argument(
        "--input",
        type=Path,
        help="Read an existing TWG JSON payload instead of calling TWG",
    )
    arguments = parser.parse_args()
    if arguments.input is None and not arguments.keys:
        parser.error("provide at least one Jira key or --input")
    if arguments.input is not None and arguments.keys:
        parser.error("Jira keys cannot be combined with --input")
    return arguments


def main():
    arguments = parse_arguments()
    try:
        payload = (
            json.loads(arguments.input.read_text(encoding="utf-8"))
            if arguments.input is not None
            else fetch_workitems(arguments.keys)
        )
        projected = [project_workitem(item) for item in normalize_payload(payload)]
    except (OSError, ValueError, RuntimeError, json.JSONDecodeError) as error:
        print(f"jira-workitem-implementation: {error}", file=sys.stderr)
        return 1
    print(json.dumps({"items": projected}, indent=2, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
